packages/xo-server/src/xo-mixins/xen-servers.js

```diff
@@ -30,6 +30,44 @@ class PoolAlreadyConnected extends BaseError {
 
 const log = createLogger('xo:xo-mixins:xen-servers')
 
+// - server is disconnected:
+//   - `_xapis[server.id]` is `undefined`
+//   - `connectXenServer()`
+//     - normal case:
+//       1. `_xapis[server.id] = new Xapi()`
+//       2. connect xapi
+//       3. `_xapis[server.id]` isn't `undefined`
+//       4. `this._serverIdsByPool[xapi.pool.$id]` is `undefined`
+//       5. `this._serverIdsByPool[xapi.pool.$id] = server.id`
+//       6. synchronize XO with xapi events
+//
+//     - request disconnection:
+//       1. `_xapis[server.id] = new Xapi()`
+//       2. connect xapi
+//       3. `_xapis[server.id]` is `undefined`
+//       4. disconnect xapi
+//
+//     - server's pool exists:
+//       1. `_xapis[server.id] = new Xapi()`
+//       2. connect xapi
+//       3. `_xapis[server.id]` isn't `undefined`
+//       4. `this._serverIdsByPool[xapi.pool.$id]` isn't `undefined`
+//       5. delete `_xapis[server.id]`
+//       6. disconnect xapi
+//       7. update the server's error
+//       8. throw an error
+
+//     - xapi disconnected by loosing the connection to the server:
+//       1. `_xapis[server.id] = new Xapi()`
+//       2. connect xapi
+//       3. `_xapis[server.id]` isn't `undefined`
+//       4. `this._serverIdsByPool[xapi.pool.$id]` is `undefined`
+//       5. `this._serverIdsByPool[xapi.pool.$id] = server.id`
+//       6. synchronize XO with xapi events
+//       7. xapi disconnected
+//       8. desynchronize XO with xapi events
+//       9. delete `_xapis[server.id]`
+//       10. delete `this._serverIdsByPool[poolId]`
```

Comment:
I don't need the whole `connectXenServer` to be rewritten as spec, I think we need to document the global state during all phases/steps:

```
Server is disconnected:
- _xapi[server.id] is undefined

Server is connecting:
- _xapi[server.id] is defined

Server is connected:
- _xapi[server.id] id defined
- _serversIdsByPool[xapi.pool.$id] is server.id
```